pkg/provisioner/crdprovisioner.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	if azVolumeCreated.Status == nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to fetch status of volume created for volume name (%s)", volumeName))
+	}
+
+	return azVolumeCreated.Status.ResponseObject, nil
+}
+
+func (c *CrdProvisioner) DeleteVolume(ctx context.Context, volumeID string, secrets map[string]string) error {
+	azV := c.azDiskClient.DiskV1alpha1().AzVolumes(c.namespace)
+	err := azV.Delete(ctx, volumeID, metav1.DeleteOptions{})
+	if errors.IsNotFound(err) {
+		klog.Info("Could not find the volume name (%s). Deletion succedded", volumeID)
+		return nil
+	}
+
+	if err != nil {
+		klog.Error("Failed to delete azvolume resource for volume id (%s), error: %v", volumeID, err)
+		return err
+	}
+
+	// Verify if the azVolume is deleted
+	azVolume, _ := azV.Get(ctx, volumeID, metav1.GetOptions{})
+	if azVolume != nil {
+		return status.Error(codes.Internal, fmt.Sprintf("Failed to delete azvolume resource for volume id (%s)", volumeID))
+	}
+
+	return nil
+}
+
+func (c *CrdProvisioner) PublishVolume(
+	ctx context.Context,
+	volumeID string,
+	nodeID string,
+	volumeCapability *v1alpha1.VolumeCapability,
+	readOnly bool,
+	secrets map[string]string,
+	volumeContext map[string]string) (*v1alpha1.PublishedVolumeParams, error) {
+	azVA := c.azDiskClient.DiskV1alpha1().AzVolumeAttachments(c.namespace)
+	azVolumeAttachment := &v1alpha1.AzVolumeAttachment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-%s", volumeID, nodeID),
+		},
+		Spec: v1alpha1.AzVolumeAttachmentSpec{
+			UnderlyingVolume: volumeID,
+			NodeName:         nodeID,
+			VolumeContext:    volumeContext,
+			RequestedRole:    v1alpha1.PrimaryRole,
+		},
+	}
+
+	attachmentCreated, err := azVA.Create(ctx, azVolumeAttachment, metav1.CreateOptions{})
+	if err != nil {
+		klog.Error("Error creating azvolume attachment for volume id (%s) to node id (%s) error : %v", volumeID, nodeID, err)
+		return nil, err
+	}
+
+	if attachmentCreated.Status == nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Failed to attach azvolume attachment resource for volume id (%s) to node (%s)", volumeID, nodeID))
+	}
+
+	return attachmentCreated.Status.ResponseObject, nil
+}
+
+func (c *CrdProvisioner) UnpublishVolume(
+	ctx context.Context,
+	volumeID string,
+	nodeID string,
+	secrets map[string]string) error {
+	azVA := c.azDiskClient.DiskV1alpha1().AzVolumeAttachments(c.namespace)
+	err := azVA.Delete(ctx, fmt.Sprintf("%s-%s", volumeID, nodeID), metav1.DeleteOptions{})
+	if errors.IsNotFound(err) {
+		klog.Info("Could not find the volume attachment (%s). Deletion succedded", volumeID)
+		return nil
+	}
+
+	if err != nil {
+		klog.Error("Failed to delete azvolume attachment resource for volume id (%s) to node (%s), error: %v", volumeID, nodeID, err)
+		return err
+	}
+
+	// Verify if the azVolume is deleted
+	azVolumeAttachment, _ := azVA.Get(ctx, fmt.Sprintf("%s-%s", volumeID, nodeID), metav1.GetOptions{})
+	if azVolumeAttachment != nil {
```

Comment:
Same goes for the deletion, `AzVolumeAttachment` will not get deleted instantaneously, as the request will first need to be dequeued from the reconciler workqueue and it will take time for the cloud provisioner's `UnpublishVolume` to complete.